File: leetcode-java/src/main/java/chapter02/OddEvenList.java
package chapter02;

import utils.ListNode;

/**
 * <p>奇偶链表</p>
 * <p>问题描述: 让结点编号为奇数的指向奇数, 为偶数的指向偶数</p>
 */
public class OddEvenList {

    /**
     * <p>思路: 采用双指针也就是分割链表的解法</p>
     */
    private static ListNode oddEvenListIndex(ListNode head){
        if (head == null || head.next == null) {
            return head;
        }
        boolean flag = true;
        ListNode current = head;
        ListNode oDummy = new ListNode(0, head), oCurrent = oDummy;
        ListNode eDummy = new ListNode(0, head.next), eCurrent = eDummy;
        while (current != null) {
            if (flag) {
                oCurrent.next = current;
                oCurrent = oCurrent.next;
            } else {
                eCurrent.next = current;
                eCurrent = eCurrent.next;
            }
            flag = !flag;
            current = current.next;
        }
        eCurrent.next = null;
        oCurrent.next = eDummy.next;
        return oDummy.next;
    }

    /**
     * <p>思路: 模拟</p>
     */
    public static ListNode oddEvenList(ListNode head){
        if (head == null || head.next == null) {
            return head;
        }
        ListNode odd = head;
        ListNode even = head.next;
        ListNode link = even;
        while (even != null) {
            if (odd.next.next == null) {
                break;
            }
            odd.next = odd.next.next;
            even.next = even.next.next;
            odd = odd.next;
            even = even.next;
        }

        odd.next = link;
        return head;
    }
}

File: leetcode-go/chapter02/odd_even_list.go
package main

type ListOddEvener struct {}

// 奇偶链表: 模拟
func (oddEvener *ListOddEvener) oddEvenList(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}
	odd, even := head, head.Next
	link := even
	for even != nil {
		if odd.Next == nil {
			break
		}
		odd.Next = odd.Next.Next
		even.Next = even.Next.Next
		odd = odd.Next
		even = even.Next
	}
	odd.Next = link
	return head
}

// 奇偶链表 => 分割链表
func (oddEvener *ListOddEvener) oddEvenListSplit(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}
	flag := true
	current := head
	oDummy, eDummy := &ListNode{Val: 0, Next: head}, &ListNode{Val: 0, Next: head.Next}
	oCurrent, eCurrent := oDummy, eDummy
	for current != nil {
		if flag {
			oCurrent.Next = current
			oCurrent = oCurrent.Next
		} else {
			eCurrent.Next = current
			eCurrent = eCurrent.Next
		}
		flag = !flag
		current = current.Next
 	}
	eCurrent.Next = nil
	oCurrent.Next = eDummy.Next
	return oDummy.Next
}
